src/cuda/gpu_cew_xc.h

```diff
@@ -0,0 +1,287 @@
+        //QUICK_SAFE_CALL((cshell_getcew_xc_grad_kernel<<< 1,1, gpu -> gpu_xcq -> smem_size>>>()));
+
+    }
+
+    cudaDeviceSynchronize();
+
+    get_cew_accdens(gpu);
+
+    prune_grid_sswgrad();
+
+    get_sswnumgrad_kernel<<< gpu->blocks, gpu->sswGradThreadsPerBlock, gpu -> gpu_xcq -> smem_size>>>();
+
+    //get_sswnumgrad_kernel<<< 1,1, gpu -> gpu_xcq -> smem_size>>>();
+
+    cudaDeviceSynchronize();
+
+    gpu_delete_sswgrad_vars();
+
+}
+
+
+void get_cew_accdens(_gpu_type gpu){
+
+    QUICKDouble *gridpt = new QUICKDouble[3]; 
+    QUICKDouble *cewGrad= new QUICKDouble[3];   
+
+    gpu -> gpu_xcq -> densa -> Download();
+
+    if(gpu -> gpu_sim.is_oshell == true) gpu -> gpu_xcq -> densb -> Download();
+    
+
+    for(int i=0; i< gpu -> gpu_xcq -> npoints;i++){
+        
+        QUICKDouble weight = gpu -> gpu_xcq -> weight -> _hostData[i];
+        QUICKDouble densea = gpu -> gpu_xcq -> densa -> _hostData[i];
+        QUICKDouble denseb = densea;
+
+        if(gpu -> gpu_sim.is_oshell == true) denseb = gpu -> gpu_xcq -> densb -> _hostData[i];   
+
+        gridpt[0] = gpu -> gpu_xcq -> gridx -> _hostData[i];
+        gridpt[1] = gpu -> gpu_xcq -> gridy -> _hostData[i];
+        gridpt[2] = gpu -> gpu_xcq -> gridz -> _hostData[i];
+
+        const QUICKDouble charge_density = -weight * (densea+denseb);
+
+        for(int j=0; j<3; j++) cewGrad[j]=0.0;
+
+        QUICKDouble const *cnst_gridpt = gridpt;
+
+#ifdef CEW 
+        // this function comes from cew library in amber       
+        cew_accdensatpt_(cnst_gridpt, &charge_density, cewGrad);
+#endif
+
+//printf("cew_accdensatpt %f %f %f %f %f %f %f \n", gridpt[0], gridpt[1], gridpt[2], charge_density\
+,cewGrad[0], cewGrad[1], cewGrad[2]);
+
+        int Istart = (gpu -> gpu_xcq -> gatm -> _hostData[i]-1) * 3;
+
+        for(int j=0; j<3; j++)
+            gpu->grad->_hostData[Istart+j] += cewGrad[j];
+
+    }
+
+    delete gridpt;
+    delete cewGrad;
+
+}
+
+
+
+
+__global__ void getcew_xc_kernel()
+{
+  unsigned int offset = blockIdx.x*blockDim.x+threadIdx.x;
+  int totalThreads = blockDim.x*gridDim.x;
+
+  for (QUICKULL gid = offset; gid < devSim_dft.npoints; gid += totalThreads) {
+
+    int bin_id    = devSim_dft.bin_locator[gid];
+    int bfloc_st  = devSim_dft.basf_locator[bin_id];
+    int bfloc_end = devSim_dft.basf_locator[bin_id+1];
+
+    QUICKDouble gridx = devSim_dft.gridx[gid];
+    QUICKDouble gridy = devSim_dft.gridy[gid];
+    QUICKDouble gridz = devSim_dft.gridz[gid];
+
+    QUICKDouble weight = devSim_dft.weight[gid];
+
+    QUICKDouble dfdr = devSim_dft.cew_vrecip[gid];
+
+    for (int i = bfloc_st; i< bfloc_end; ++i) {
+
+      int ibas = devSim_dft.basf[i];
+      QUICKDouble phi, dphidx, dphidy, dphidz;
+
+      pteval_new(gridx, gridy, gridz, &phi, &dphidx, &dphidy, &dphidz, devSim_dft.primf, devSim_dft.primf_locator, ibas, i);
```

Comment:
In principle we only need the basis function values here, not gradients. We could write another `pteval_new()` function that only returns the basis function values. But I don't know how to change the threshold for DMCutoff accordingly...

Reply:
I left this as it is since the kernel time is very small and gradients are still required for cutoff criteria.

Reply:
OK, thanks.